src/Features/CSharp/Portable/CodeRefactorings/PushMemberUp/ClassPusher.cs

```diff
@@ -0,0 +1,109 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.  
+
+using System.Linq;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Editing;
+using static Microsoft.CodeAnalysis.CodeActions.CodeAction;
+
+namespace Microsoft.CodeAnalysis.CSharp.CodeRefactorings.PushMemberUp
+{
+    internal class ClassPusher : AbstractMemberPusher
+    {
+        internal ClassPusher(INamedTypeSymbol targetClassSymbol, SemanticModel semanticModel, SyntaxNode userSelectedNode, Document contextDocument):
+            base(targetClassSymbol, semanticModel, userSelectedNode, contextDocument)
+        {
+        }
+
+        protected override bool AreModifiersValid(SyntaxNode userSelectedNode)
+        {
+            // Abstract member can't be push to non-abstract class
+            if (TargetTypeSymbol.IsAbstract)
+            {
+                return true;
+            }
+            else
+            {
+                return !SemanticModel.GetDeclaredSymbol(userSelectedNode).IsAbstract;
+            }
+        }
+
+        protected override CodeAction CreateDocumentChangeAction(MemberDeclarationSyntax memberToPush, Document contextDocument)
+        {
+            return new DocumentChangeAction(
+                Title,
+                async _ =>
```

Comment:
in general, we make thse lambdas small.  just call to an async function.  it makes it easier to understand, and makes sure thi lambda doesn't capture data it shouldn't unintentionally.